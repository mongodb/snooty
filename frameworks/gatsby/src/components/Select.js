import React, { useRef, forwardRef } from 'react';
import { cx, css } from '@leafygreen-ui/emotion';
import styled from '@emotion/styled';
import { Option, Select as LGSelect } from '@leafygreen-ui/select';
import PropTypes from 'prop-types';
import { palette } from '@leafygreen-ui/palette';
import { color, focusRing } from '@leafygreen-ui/tokens';
import Icon from '@leafygreen-ui/icon';
import { theme } from '../theme/docsTheme';
import { isOfflineDocsBuild } from '../utils/is-offline-docs-build';

const Label = styled('p')`
  font-size: ${theme.fontSize.small};
  line-height: ${theme.fontSize.default};
  font-weight: bolder;
  letter-spacing: 0;
  /* TODO: Remove !important when mongodb-docs.css is removed */
  margin: 0 0 12px !important;
`;

const portalStyle = css`
  position: relative;
`;

const iconStyle = css`
  display: inline-block;
  margin-right: ${theme.size.small};
  max-height: 20px;
  width: 30px;
`;

/* Override LG mobile style of enlarged mobile font */
const selectWrapperStyle = css`
  @media ${theme.screenSize.upToLarge} {
    label,
    p,
    button,
    div,
    span {
      font-size: ${theme.fontSize.small};
    }
  }
`;

const labelStyle = css`
  label {
    color: ${color.light.text.primary.default};

    .dark-theme & {
      color: ${color.dark.text.primary.default};
    }
  }
`;

const disabledLabelStyle = css`
  label {
    color: ${color.light.text.disabled.default};

    .dark-theme & {
      color: ${color.dark.text.disabled.default};
    }
  }
`;

const selectStyle = css`
  > button {
    background-color: ${color.light.background.primary.default};
    text-align: left;
    ${isOfflineDocsBuild && `color: ${palette.black}`};

    // Override button default color
    > *:last-child {
      > svg {
        color: ${color.light.icon.primary.default};
      }
    }

    &:focus-visible {
      box-shadow: ${focusRing.light.input};
      border-color: rgba(255, 255, 255, 0);
    }

    .dark-theme & {
      border-color: ${color.dark.border.primary.default};
      background-color: ${palette.gray.dark4};

      // Override button default color
      > *:last-child {
        > svg {
          color: ${color.dark.icon.primary.default};
        }
      }
    }
  }
`;

const enabledSelectHoverStyles = css`
  > button {
    .dark-theme & {
      &:hover,
      &:active,
      &:focus {
        background-color: ${palette.gray.dark4};
        color: ${color.dark.text.primary.hover};
      }

      &:focus-visible {
        background-color: ${palette.gray.dark4};
        box-shadow: ${focusRing.dark.input};
        border-color: rgba(255, 255, 255, 0);
      }
    }
  }
`;

const disabledSelectStyles = css`
  > button {
    cursor: not-allowed;
    pointer-events: unset;
    box-shadow: unset;

    &:active {
      pointer-events: none;
    }

    &[aria-disabled='true'] {
      background-color: ${color.light.background.disabled.default};
      border-color: ${color.light.border.disabled.default};

      .dark-theme & {
        background-color: ${color.dark.background.disabled.default};
        border-color: ${color.dark.border.disabled.default};
      }

      &:hover,
      &:active {
        box-shadow: inherit;
      }

      > *:last-child {
        > svg {
          color: ${color.light.icon.disabled.default};

          .dark-theme & {
            color: ${color.dark.icon.disabled.default};
          }
        }
      }
    }
  }
`;

const optionStyling = css`
  align-items: center;

  & > span > svg {
    display: none;
  }
`;

const OFFLINE_SELECT_ID = 'offline-select';

const PortalContainer = forwardRef(({ ...props }, ref) => (
  <div id={isOfflineDocsBuild ? OFFLINE_SELECT_ID : null} className={cx(portalStyle, props.className)} ref={ref}>
    {props.children}
  </div>
));

const offlineMenuStyling = css`
  position: absolute;
  background: white;
  font-size: 13px;
  width: 100%;
`;

const offlineListStyle = css`
  list-style: none;
  padding: 8px 0px;
  border-radius: 12px;
  box-shadow: rgba(0, 30, 43, 0.25) 0px 4px 7px 0px;
  margin: 6px 0 0 0;
  max-height: 250px;
  overflow: scroll;
`;

const offlineListItemStyle = css`
  display: flex;
  width: 100%;
  outline: none;
  overflow-wrap: anywhere;
  position: relative;
  padding: 8px 12px;
  cursor: pointer;
  color: ${palette.gray.dark3};
  align-items: center;
  line-height: ${theme.fontSize.default};

  &:hover {
    background-color: ${palette.gray.light2};
  }

  svg {
    opacity: 0;
    margin-right: 6px;
  }

  &[selected='true'] {
    font-weight: bold;
    svg {
      opacity: 1;
    }
  }
`;

const OfflineMenu = ({ choices }) => {
  return (
    <div className={cx(offlineMenuStyling, 'offline-select-menu')}>
      <ul className={cx(offlineListStyle)}>
        {choices.map((choice, idx) => (
          <li
            className={cx('offline-select-choice', offlineListItemStyle)}
            data-value={choice.value}
            data-text={choice.text}
            key={idx}
          >
            <Icon fill={palette.blue.base} glyph={'Checkmark'} />
            <span>{choice.text}</span>
          </li>
        ))}
      </ul>
    </div>
  );
};

const Select = ({
  className,
  choices,
  onChange,
  usePortal = true,
  defaultText = '',
  disabled = false,
  label = null,
  value = null,
  ...props
}) => {
  // show select after portal container has loaded for scroll + zindex consistency
  const portalContainer = useRef();

  return (
    <PortalContainer
      className={`${className} ${cx(selectWrapperStyle, labelStyle, { [disabledLabelStyle]: disabled })}`}
      ref={portalContainer}
    >
      <LGSelect
        data-testid="lg-select"
        value={value || ''}
        label={label}
        aria-labelledby={(!label && 'select') || null}
        size="default"
        allowDeselect={false}
        disabled={disabled}
        usePortal={usePortal}
        portalContainer={portalContainer.current}
        scrollContainer={portalContainer.current}
        popoverZIndex={2}
        placeholder={defaultText}
        defaultValue={value ? String(value) : ''}
        onChange={(value) => {
          onChange({ value });
        }}
        className={cx(selectStyle, disabled ? disabledSelectStyles : enabledSelectHoverStyles)}
        {...props}
      >
        {choices.map((choice) => (
          <Option
            className={cx(optionStyling)}
            key={choice.value}
            value={choice.value}
            glyph={choice.icon}
            role="option"
          >
            {choice.tabSelectorIcon && <choice.tabSelectorIcon className={cx(iconStyle)} />}
            {choice.text}
          </Option>
        ))}
      </LGSelect>
      {isOfflineDocsBuild && <OfflineMenu choices={choices} />}
    </PortalContainer>
  );
};

Select.propTypes = {
  choices: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  ).isRequired,
  usePortal: PropTypes.bool,
  defaultText: PropTypes.string,
  disabled: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  label: PropTypes.string,
  value: PropTypes.string,
};

Select.defaultProps = {
  usePortal: true,
  defaultText: '',
  disabled: false,
  label: null,
  value: null,
};

export { Select as default, Label };
