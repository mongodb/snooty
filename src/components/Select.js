import React, { useCallback, useEffect, useMemo, useState } from 'react';
import { css } from '@emotion/core';
import styled from '@emotion/styled';
import Icon from '@leafygreen-ui/icon';
import PropTypes from 'prop-types';
import { theme } from '../theme/docsTheme';

const BORDER_SIZE = '1px';
const ENTER_KEY = 13;
const ESCAPE_KEY = 27;
// This is the height of the closed select
const OPTIONS_POSITION_OFFSET = '36px';

const activeSelectStyles = css`
  border-bottom: none;
  /* Remove bottom border radius for seamless border transition */
  border-radius: 2px 2px 0 0;
  box-shadow: none;
  /* Remove the focus outline due to bottom being cut off by expanded element */
  :focus {
    outline: none;
  }
`;

const Label = styled('p')`
  font-weight: bolder;
  letter-spacing: 0;
  margin: 0 0 ${theme.size.tiny};
`;

const SelectedText = styled('p')`
  margin: 0;
`;

const Option = styled('li')`
  align-items: center;
  background-color: #fff;
  color: black;
  display: flex;
  overflow: hidden;
  padding: 10px ${theme.size.default};
  text-overflow: ellipsis;
  white-space: nowrap;
  :focus,
  :hover {
    background-color: rgba(231, 238, 236, 0.2);
  }
`;

const Options = styled('ul')`
  border: ${BORDER_SIZE} solid rgba(184, 196, 194, 0.4);
  border-radius: 0 0 2px 2px;
  box-shadow: 0 0 ${theme.size.tiny} 0 rgba(233, 233, 233, 0.3);
  left: -${BORDER_SIZE};
  padding: 0;
  position: absolute;
  margin: 0;
  top: ${OPTIONS_POSITION_OFFSET};
  /* account for border */
  width: calc(100% + 2px);
  z-index: 5;
`;

const SelectContainer = styled('div')`
  font-size: ${theme.fontSize.small};
  letter-spacing: 0.5px;
  line-height: ${theme.size.default};
  opacity: ${({ enabled }) => (enabled ? 1 : 0.3)};
`;

const StyledCustomSelect = styled('div')`
  background-color: #fff;
  border: ${BORDER_SIZE} solid rgba(184, 196, 194, 0.4);
  border-radius: 2px;
  box-shadow: 0 0 ${theme.size.tiny} 0 rgba(233, 233, 233, 0.3);
  color: #000;
  cursor: ${({ enabled }) => (enabled ? 'pointer' : 'not-allowed')};
  position: relative;
  ${({ showOptions }) => showOptions && activeSelectStyles};
`;

const SelectedOption = styled('div')`
  align-items: center;
  background-color: #fff;
  display: flex;
  justify-content: space-between;
  padding: 10px ${theme.size.default};
  position: relative;
`;

const Select = ({ choices, onChange, defaultText = '', disabled = false, label = null, value = null, ...props }) => {
  const enabled = useMemo(() => !disabled, [disabled]);
  const [selectText, setSelectText] = useState(defaultText);
  const [showOptions, setShowOptions] = useState(false);
  const toggleSelectExpand = useCallback(() => {
    if (enabled) setShowOptions(!showOptions);
  }, [enabled, showOptions]);
  /**
   * This useEffect should only be called once the component first renders with choices,
   * this should populate the select item with the default choice if there is one
   */
  useEffect(() => {
    if (choices.length && value) {
      const choice = choices.filter(choice => choice.value === value);
      if (choice && choice.length) {
        const currentChoice = choice[0];
        setSelectText(currentChoice.text);
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [choices]);
  const showOptionsOnEnter = useCallback(
    e => {
      if (e.keyCode === ENTER_KEY) {
        toggleSelectExpand();
      } else if (e.keyCode === ESCAPE_KEY && showOptions) {
        // Hitting the escape key should only close the select
        toggleSelectExpand();
      }
    },
    [showOptions, toggleSelectExpand]
  );

  const selectOption = useCallback(
    choice => {
      onChange(choice);
      setSelectText(choice.text);
      setShowOptions(false);
    },
    [onChange]
  );

  const optionOnEnter = useCallback(
    option => e => {
      if (e.keyCode === ENTER_KEY) {
        selectOption(option);
      }
    },
    [selectOption]
  );

  const closeOptionsOnBlur = useCallback(
    e => {
      // Check the event to see if the next element would be a list element
      // otherwise, close the options
      const isTabbingThroughOptions = e.relatedTarget && e.relatedTarget.tagName === 'LI';
      if (!isTabbingThroughOptions) {
        setShowOptions(false);
      }
    },
    [setShowOptions]
  );
  return (
    <SelectContainer aria-disabled={disabled} enabled={enabled} {...props}>
      {label && <Label>{label}</Label>}
      <StyledCustomSelect
        aria-expanded={showOptions}
        enabled={enabled}
        onBlur={closeOptionsOnBlur}
        onClick={toggleSelectExpand}
        onKeyDown={showOptionsOnEnter}
        role="listbox"
        showOptions={showOptions}
        tabIndex="0"
      >
        <SelectedOption showOptions={showOptions} role="option">
          <SelectedText>{selectText}</SelectedText>
          <Icon glyph={showOptions ? 'CaretUp' : 'CaretDown'} />
        </SelectedOption>
        {showOptions && (
          <Options>
            {choices.map(({ text, value }) => (
              <Option
                key={value}
                onClick={() => selectOption({ text, value })}
                onKeyDown={optionOnEnter({ text, value })}
                role="option"
                tabIndex="0"
              >
                {text}
              </Option>
            ))}
          </Options>
        )}
      </StyledCustomSelect>
    </SelectContainer>
  );
};

Select.propTypes = {
  choices: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  ).isRequired,
  defaultText: PropTypes.string,
  disabled: PropTypes.bool,
  onChange: PropTypes.func.isRequired,
  label: PropTypes.string,
  value: PropTypes.string,
};

Select.defaultProps = {
  defaultText: '',
  disabled: false,
  label: null,
  value: null,
};

export default Select;
