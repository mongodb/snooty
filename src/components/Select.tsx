import React, { useRef, forwardRef, ReactNode } from 'react';
import { cx, css } from '@leafygreen-ui/emotion';
import styled from '@emotion/styled';
import { Option, Select as LGSelect } from '@leafygreen-ui/select';
import { palette } from '@leafygreen-ui/palette';
import { color, focusRing } from '@leafygreen-ui/tokens';
import Icon, { LGGlyph } from '@leafygreen-ui/icon';
import { theme } from '../theme/docsTheme';
import { isOfflineDocsBuild } from '../utils/is-offline-docs-build';

const Label = styled('p')`
  font-size: ${theme.fontSize.small};
  line-height: ${theme.fontSize.default};
  font-weight: bolder;
  letter-spacing: 0;
  /* TODO: Remove !important when mongodb-docs.css is removed */
  margin: 0 0 12px !important;
`;

const portalStyle = css`
  position: relative;
`;

const iconStyle = css`
  display: inline-block;
  margin-right: ${theme.size.small};
  max-height: 20px;
  width: 30px;
`;

/* Override LG mobile style of enlarged mobile font */
const selectWrapperStyle = css`
  @media ${theme.screenSize.upToLarge} {
    label,
    p,
    button,
    div,
    span {
      font-size: ${theme.fontSize.small};
    }
  }
`;

const labelStyle = css`
  label {
    color: ${color.light.text.primary.default};

    .dark-theme & {
      color: ${color.dark.text.primary.default};
    }
  }
`;

const disabledLabelStyle = css`
  label {
    color: ${color.light.text.disabled.default};

    .dark-theme & {
      color: ${color.dark.text.disabled.default};
    }
  }
`;

const selectStyle = css`
  > button {
    background-color: ${color.light.background.primary.default};
    text-align: left;
    ${isOfflineDocsBuild && `color: ${palette.black}`};

    // Override button default color
    > *:last-child {
      > svg {
        color: ${color.light.icon.primary.default};
      }
    }

    &:focus-visible {
      box-shadow: ${focusRing.light.input};
      border-color: rgba(255, 255, 255, 0);
    }

    .dark-theme & {
      border-color: ${color.dark.border.primary.default};
      background-color: ${palette.gray.dark4};

      // Override button default color
      > *:last-child {
        > svg {
          color: ${color.dark.icon.primary.default};
        }
      }
    }
  }
`;

const enabledSelectHoverStyles = css`
  > button {
    .dark-theme & {
      &:hover,
      &:active,
      &:focus {
        background-color: ${palette.gray.dark4};
        color: ${color.dark.text.primary.hover};
      }

      &:focus-visible {
        background-color: ${palette.gray.dark4};
        box-shadow: ${focusRing.dark.input};
        border-color: rgba(255, 255, 255, 0);
      }
    }
  }
`;

const disabledSelectStyles = css`
  > button {
    cursor: not-allowed;
    pointer-events: unset;
    box-shadow: unset;

    &:active {
      pointer-events: none;
    }

    &[aria-disabled='true'] {
      background-color: ${color.light.background.disabled.default};
      border-color: ${color.light.border.disabled.default};

      .dark-theme & {
        background-color: ${color.dark.background.disabled.default};
        border-color: ${color.dark.border.disabled.default};
      }

      &:hover,
      &:active {
        box-shadow: inherit;
      }

      > *:last-child {
        > svg {
          color: ${color.light.icon.disabled.default};

          .dark-theme & {
            color: ${color.dark.icon.disabled.default};
          }
        }
      }
    }
  }
`;

const optionStyling = css`
  align-items: center;

  & > span > svg {
    display: none;
  }
`;

const OFFLINE_SELECT_ID = 'offline-select';

const PortalContainer = forwardRef<HTMLDivElement, { className?: string; children: ReactNode }>(({ ...props }, ref) => (
  <div id={isOfflineDocsBuild ? OFFLINE_SELECT_ID : undefined} className={cx(portalStyle, props.className)} ref={ref}>
    {props.children}
  </div>
));

const offlineMenuStyling = css`
  position: absolute;
  background: white;
  font-size: 13px;
  width: 100%;
`;

const offlineListStyle = css`
  list-style: none;
  padding: 8px 0px;
  border-radius: 12px;
  box-shadow: rgba(0, 30, 43, 0.25) 0px 4px 7px 0px;
  margin: 6px 0 0 0;
  max-height: 250px;
`;

const offlineListItemStyle = css`
  display: flex;
  width: 100%;
  outline: none;
  overflow-wrap: anywhere;
  position: relative;
  padding: 8px 12px;
  cursor: pointer;
  color: ${palette.gray.dark3};
  align-items: center;
  line-height: ${theme.fontSize.default};

  &:hover {
    background-color: ${palette.gray.light2};
  }

  svg {
    opacity: 0;
    margin-right: 6px;
  }

  &[selected='true'] {
    font-weight: bold;
    svg {
      opacity: 1;
    }
  }
`;

export const OfflineMenu = ({ choices, className }: { choices: Array<SelectOption>; className?: string }) => {
  return (
    <div className={cx(offlineMenuStyling, 'offline-select-menu', className)}>
      <ul className={cx(offlineListStyle)}>
        {choices.map((choice, idx) => (
          <li
            className={cx('offline-select-choice', offlineListItemStyle)}
            data-value={choice.value}
            data-text={choice.text}
            key={idx}
          >
            <Icon fill={palette.blue.base} glyph={'Checkmark'} />
            <span>{choice.text}</span>
          </li>
        ))}
      </ul>
    </div>
  );
};

export type SelectOption = {
  text: string;
  value: string;
  icon?: LGGlyph.Element;
  tabSelectorIcon?: React.ComponentType<{ className?: string }>;
};

export type SelectProps = {
  className?: string;
  choices: Array<SelectOption>;
  onChange: ({ value }: { value: string }) => void;
  usePortal?: boolean;
  defaultText?: string;
  disabled?: boolean;
  label?: string;
  value?: string;
};

const Select = ({
  className,
  choices,
  onChange,
  usePortal = true,
  defaultText = '',
  disabled = false,
  label,
  value = '',
  ...props
}: SelectProps) => {
  // show select after portal container has loaded for scroll + zindex consistency
  const portalContainer = useRef<HTMLDivElement>(null);

  return (
    <PortalContainer
      className={`${className} ${cx(selectWrapperStyle, labelStyle, { [disabledLabelStyle]: disabled })}`}
      ref={portalContainer}
    >
      <LGSelect
        data-testid="lg-select"
        value={value}
        // Only one is allowed (label or aria-labelledby)
        {...(label ? { label } : { 'aria-labelledby': 'select' })}
        size="default"
        allowDeselect={false}
        disabled={disabled}
        usePortal={usePortal}
        portalContainer={portalContainer.current}
        scrollContainer={portalContainer.current}
        popoverZIndex={2}
        placeholder={defaultText}
        onChange={(value) => {
          onChange({ value });
        }}
        className={cx(selectStyle, disabled ? disabledSelectStyles : enabledSelectHoverStyles)}
        {...props}
      >
        {choices.map((choice) => (
          <Option
            className={cx(optionStyling)}
            key={choice.value}
            value={choice.value}
            glyph={choice.icon}
            {...({ role: 'option' } as React.HTMLAttributes<HTMLLIElement>)}
          >
            {choice.tabSelectorIcon && <choice.tabSelectorIcon className={cx(iconStyle)} />}
            {choice.text}
          </Option>
        ))}
      </LGSelect>
      {isOfflineDocsBuild && <OfflineMenu choices={choices} />}
    </PortalContainer>
  );
};

export { Select as default, Label };
