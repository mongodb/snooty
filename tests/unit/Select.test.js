import React, { useState } from 'react';
import { shallow, mount } from 'enzyme';
import Select from '../../src/components/Select';

const DEFAULT_ATLAS_CHOICE = { text: 'MongoDB Atlas', value: 'atlas' };
const DEFAULT_SERVER_CHOICE = { text: 'MongoDB Server', value: 'server' };
const DEFAULT_CHOICES = [DEFAULT_SERVER_CHOICE, DEFAULT_ATLAS_CHOICE];

// Simple wrapper to add state control around the Pagination component
const SelectController = ({
  choices = DEFAULT_CHOICES,
  customOnChange = null,
  defaultText = '',
  disabled = false,
  label = null,
  value = null,
}) => {
  const [selectValue, setSelectValue] = useState(value);
  const onChange = choice => {
    if (customOnChange) customOnChange(choice);
    setSelectValue(choice.value);
  };
  return (
    <Select
      choices={choices}
      onChange={onChange}
      defaultText={defaultText}
      disabled={disabled}
      label={label}
      value={selectValue}
    />
  );
};

describe('Select', () => {
  // Helper to open the dropdown passed a series of simulate args (click, keypress)
  const dropdownOpen = (args = ['click'], props = {}) => {
    const wrapper = mount(<SelectController {...props} />);
    // Dropdown should be closed by default
    expect(!wrapper.find('Options').exists());
    const selectParent = wrapper.find('StyledCustomSelect');
    selectParent.simulate(...args);
    // Now it should be open
    expect(wrapper.find('Options').exists());
    return wrapper;
  };

  it('renders select correctly', () => {
    const wrapper = shallow(<SelectController />);
    expect(wrapper).toMatchSnapshot();
  });

  it('displays default text', () => {
    const defaultText = 'Some default text';
    const wrapper = mount(<SelectController defaultText={defaultText} />);
    const renderedText = wrapper.find('SelectedText').text();
    expect(renderedText).toBe(defaultText);
  });

  it('conditionally should render a label', () => {
    const wrapperWithoutLabel = mount(<SelectController />);
    expect(!wrapperWithoutLabel.find('Label').exists());
    const labelText = 'Select Label';
    const wrapperWithLabel = mount(<SelectController label={labelText} />);
    expect(wrapperWithLabel.find('Label').text()).toBe(labelText);
  });

  it('opens a dropdown with options when clicked', () => {
    dropdownOpen();
  });

  it('opens a dropdown with options with the enter key for accessibility', () => {
    dropdownOpen(['keypress', { key: 'Enter' }]);
  });

  it('prevents interaction when disabled', () => {
    const wrapper = mount(<SelectController disabled />);
    // Dropdown should be closed by default
    expect(!wrapper.find('Options').exists());
    const selectParent = wrapper.find('StyledCustomSelect');
    selectParent.simulate('click');
    selectParent.simulate('keypress', { key: 'Enter' });
    // It should still be closed
    expect(!wrapper.find('Options').exists());
  });

  it('closes the dropdown by clicking again on the toggle parent', () => {
    const wrapper = dropdownOpen();
    const selectParent = wrapper.find('StyledCustomSelect');
    selectParent.simulate('click');
    // Dropdown was previously open, it should now be closed
    expect(!wrapper.find('Options').exists());
  });

  it('updates the selected text when an item is clicked', () => {
    const defaultText = 'Default Text';
    const wrapper = dropdownOpen(['click'], { defaultText });
    let renderedText = wrapper.find('SelectedText').text();
    expect(renderedText).toBe(defaultText);
    const firstOption = wrapper.find('Option').at(0);
    expect(firstOption.text()).toBe(DEFAULT_CHOICES[0].text);
    firstOption.simulate('click');
    renderedText = wrapper.find('SelectedText').text();
    expect(renderedText).toBe(DEFAULT_CHOICES[0].text);
  });

  it('passes the entire choice to the onChange callback', () => {
    const customOnChange = jest.fn();
    const wrapper = dropdownOpen(['click'], { customOnChange });
    const firstOption = wrapper.find('Option').at(0);
    expect(firstOption.text()).toBe(DEFAULT_CHOICES[0].text);
    firstOption.simulate('click');
    expect(customOnChange.mock.calls.length).toBe(1);
    expect(customOnChange.mock.calls[0][0]).toStrictEqual(DEFAULT_CHOICES[0]);
  });

  it('should update selected text given a value', () => {
    const wrapper = mount(<SelectController value={DEFAULT_CHOICES[0].value} />);
    expect(wrapper.find('SelectedText').text()).toBe(DEFAULT_CHOICES[0].text);
  });

  it('should reset the form when null/empty string is passed as value', () => {
    const defaultText = 'Reset text';
    // Bypass SelectController so we can directly modify the value prop
    const wrapper = mount(
      <Select
        defaultText={defaultText}
        choices={DEFAULT_CHOICES}
        onChange={() => {}}
        value={DEFAULT_CHOICES[0].value}
      />
    );
    expect(wrapper.find('SelectedText').text()).toBe(DEFAULT_CHOICES[0].text);
    wrapper.setProps({ value: '' });
    expect(wrapper.find('SelectedText').text()).toBe('Reset text');
  });
});
