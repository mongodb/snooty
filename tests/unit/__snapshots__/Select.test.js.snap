// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders select correctly 1`] = `
<Select
  choices={
    Array [
      Object {
        "text": "MongoDB Server",
        "value": "server",
      },
      Object {
        "text": "MongoDB Atlas",
        "value": "atlas",
      },
    ]
  }
  defaultText=""
  disabled={false}
  label={null}
  onChange={[Function]}
  value={null}
/>
`;
